README.md
[![NPM](https://nodei.co/npm/yocto-mongoose.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/yocto-mongoose/)

![alt text](https://david-dm.org/yoctore/yocto-mongoose.svg "Dependencies Status")
[![Code Climate](https://codeclimate.com/github/yoctore/yocto-mongoose/badges/gpa.svg)](https://codeclimate.com/github/yoctore/yocto-mongoose)
[![Test Coverage](https://codeclimate.com/github/yoctore/yocto-mongoose/badges/coverage.svg)](https://codeclimate.com/github/yoctore/yocto-mongoose/coverage)
[![Issue Count](https://codeclimate.com/github/yoctore/yocto-mongoose/badges/issue_count.svg)](https://codeclimate.com/github/yoctore/yocto-mongoose)
[![Build Status](https://travis-ci.org/yoctore/yocto-mongoose.svg?branch=master)](https://travis-ci.org/yoctore/yocto-mongoose)

## Overview

This module is a part of yocto node modules for NodeJS.

Please see [our NPM repository](https://www.npmjs.com/~yocto) for complete list of available tools (completed day after day).

This module provide a simple database connector for an app based on MongoDb.
Extra module can be use directly from MongoDb to Redis and/or Elasticsearch if needed.

## Motivation

Create an easy and ready to use connector & model builder based on mongoose / redis / Elasticsearch.

## Folder structure

A specific folder structure was setup for database configuration :

- enums : Contains all enums defined for database
- models : Contains all model definition for an automatic build
- validators : Contains all validator function defined on model configuration files
- methods : Contains all methods defined on model configuration

All of these folders must be set up before any usage.
For sure it's possible to set your own directory for each items. (See How to setup path directory part)

## Dependencies & Validator

- Validate rules **MUST** be build with [joi](https://www.npmjs.com/package/joi) package
- Promise was build with [Q](https://www.npmjs.com/package/q) package

## Model type & list defintiion

- String
- ObjectId
- Boolean
- Object
- Array or []

## How to set directories path

A method was defined for each path

- models(path) : set directory for models files
- validators(path) : set directory for validators files
- methods(path) : set directory for methods files
- enums(path) : set directory for enums files

## How to define a new model

For this example we want to define "Notify" model.

> First you need to go in your models directory
> Create your own folder (not mandatory)
> Create a json file in created folder named like your model name (notify.json)
> Add you model definition on it (see below for example)

```javascript
{
  "model" : {
    "name" : "Notify", // !!! IMPORANT !!! => This is your model name.
    "crud" : { // !!! MANDATORY !!! you need define it all the time
      "enable"  : true, // Set to false if you d'ont need enable automatic add of crud method
      "exclude" : [] // define here witch method we want to exlude (see CRUD parts for available method)
    },
    "properties" : { // Define here your property
      "status" : {
        "required" : true,
        "type"     : "String"
      },
      "notify_type" : {
        "required" : true,
        "type"     : "String"
      }
    }
  }
}
```

> Save file, Build & Use (See How To use part)

## How to define a validator / function / enums  attached to the previous define model

### How to add a new validator

> First edit you notify.json file created before and set the validator properties by the name of needed function, for example my Function name is "testValidator"


```javascript
{
  "model" : {
    "name" : "Notify",
     .....
    "validator" : "testValidator",
  }
}
```

> Create your structure of directory in validators directory (not mandatory).
> Create your notify.js files that will be contain your validation rules.
> Add into this files your new validation rules

```javascript
'use strict';

var joi = require('joi');

// valid account schema
exports.testValidator = function(data) {
  return joi.object().keys({
    a : joi.string()
  });
};
```

> Save file, Build & Use (See How To use part)

### How to add new function(s)

> First edit your notify.json file and add an "fn" properties, and add list of your methods :

```javascript
{
  "model" : {
    "name" : "Notify",
     .....
    "validator" : "testValidator",
    "fn"        : [
      {
        "type"  : "static", // create a static method on schema
        "name"  : "test1"
      },
      {
        "type"  : "method", // create a instance method on schema
        "name"  : "test2"
      }
  }
}
```

> Create your structure of directory in methods directory (not mandatory).

```javascript
// test 1
exports.test1 = function() {
    console.log('test1')
};
// test2
exports.test2 = function() {
  console.log('test2');
};
```
> Save file, Build & Use (See How To use part)

### How to access to mongooseTypes from custom functions

```javascript
exports.testTypes = function (enums) {
  var fn = db.getModel('Model');
  console.log(fn.Types);
}
```

**!!! IMPORTANT !!!** Each function is executed in schema context, so crud or mongoose method was also available.

### How to add enum value(s)

> Create a "file.json" into enums directory

> Add item like this :

```javascript
[
  {
    "name"  : "notify_type_list",
    "value" : [ "email", "sms", "notification" ]
  },
  {
    "name"  : "notify_status_list",
    "value" : [ "pending", "processed", "error" ]
  }
]
```

> Use it on your validator by injection dependencices to retreive a value :

```javascript
exports.testValidator = function (enums) {
  var status  = enums().get('notify_status_list');
  var type    = enums().get('notify_type_list');
 // your code here
}
```

### How to access to mongooseTypes from enums

```javascript
exports.testTypes = function (enums) {
 // your code here
 console.log(enums().Types);
}
```

## CRUD methods

All defined model have CRUD Method attached by mongoose static function.

See Below list of method & aliases :

| Operation        | Available Method | Alias(es)       | Mongoose methods                             |
|------------------|------------------|-----------------|----------------------------------------------|
| Create           | create           | insert          | save                                         |
| Read (Retrieve)  | get / getOne     | read / readOne  | find / findById / findOne                    |
| Update (Modify   | update           | modify          | findByIdAndUpdate / findOneAndUpdate / where |
| Delete (Destroy) | delete           | destroy         | findByIdAndRemove                            |

## Elasticsearch implementation

### Setting up one host or multiple hosts

You can provide to our package a multiple hosts connection for `elasticsearch` part.
A method `elasticHosts` are available to define which hosts to use on `mongoosastic`.

```javascript

var db = require('yocto-mongoose')();

// Connect
db.connect('MONGO_URL').then(function() {
  db.enableElasticsearch([ { host : '127.0.0.1', port : 9200 } ]);
});

```

### Basic configuration

Elastic search implemantion are builed with [mongoosastic](https://www.npmjs.com/package/mongoosastic) package

Elastic search rules are setted up during model build.
You can define index directly on model definition. See a simple example below :

```javascript
{
  "model" : {
    "name" :
    "crud" : {
      "enable"  : true,
      "exclude" : []
    },
    "elastic" : {
      "enable" : true, // !!! IMPORTANT !!! must be provide to enable elastic search mapping
      "options" : {} // see : https://github.com/mongoosastic/mongoosastic#setup (host, hosts, protocol, port is removed if given : See multiple host usage)
    },
    "properties" : {
      "status" : {
        "required" : true,
        "type"     : "String"
        "es_indexed" : true // here we define that this properties must be indexed.
        "es_type"    : "String" // another possible value
        "es_include_in_parent" : true // same thing
      },
      "notify_type" : {
        "required" : true,
        "type"     : "String"
      }
    }
  }
}
```

### Usage

On each model, where elastic is enabled a method search was added.

To do an `elasticsearch` query just do :

```javascript

  var query = {}; // Here define your query
  var hydrate = false; // set to true if you wan use hydrate method
  var hydrateOptions = {}; // set here your hydrate options

  // process request
  account.esearch(query, hydrate, hydrateOptions).then(function(success) {
    // your logic code here
  }).catch(function (error) {
    // your logic code here
  });
```

### List of available keys

For more complex implementation see [mongoosastic](https://www.npmjs.com/package/mongoosastic) package documentation to see available keys

### Use ssl connection or other options

```javascript

var db = require('yocto-mongoose')();

// define here your elasticsearch options
var options = {
  ssl : {
    ca: fs.readFileSync(__dirname + "/cert.pem"),
    rejectUnauthorized: true
  }
};

// Connect
db.connect('MONGO_URL').then(function() {
  db.elasticHosts([ { host : '127.0.0.1', port : 9200, protocol : 'https' } ], options);
});

```

To discover a complete list of options see : [official elastic configuration](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/configuration.html)

## Redis implementation

Our redis implementation use ioredis module but only use set/get method with expire time for the moment.

Be careful : Redis > 2.6 is required to use this module. In fact expire time is not implemented on Redis < 2.6.
For more details please read [official redis documentation](http://redis.io/commands/set)

### Setting up one host or multiple hosts

By default redis instance use single connection :

```javascript

var db = require('yocto-mongoose')();

// Connect
db.connect('MONGO_URL').then(function() {
  db.enableRedis([ { host : '127.0.0.1', port : 6379 }, { host : '127.0.0.1', port : 6379 }]);
});

```

### Setting up one host or multiple host in a cluster mode

Just add true on second parameters of `enableRedis` method.

```javascript

var db = require('yocto-mongoose')();

// Connect
db.connect('MONGO_URL').then(function() {
  db.enableRedis([ { host : '127.0.0.1', port : 6379 }, { host : '127.0.0.1', port : 6379 }], true);
});

```

### Usage

Redis is implemented in two ways in this modules, from custom methods and crud methods.

#### Redis on Crud methods

To use redis on Crud method just extend your config file like this :

```javascript
{
  "model" : {
    "name" : "Notify",
    "crud" : {
      "enable"  : true,
      "exclude" : [],
      "redis"   : { // HERE YOUR REDIS PART
        "enable"  : true, // MUST BE SET TO TRUE TO ENABLE THIS FUNCTIONALITY
        "expire"  : 10, // EXPIRE TIME FOR STORAGE IN SECONDS
        "include" : [ "get", "getOne" ] // CRUD METHODS ALLOWED TO USE REDIS
      }
    },
    "properties" : {
      "status" : {
        "required" : true,
        "type"     : "String"
      },
      "notify_type" : {
        "required" : true,
        "type"     : "String"
      }
    }
  }
}
```

This configuration enable to your current model redis on `include` method (get or getOne).
On each request on `include` method redis will be check if value exists before each mongo request.
In case of value was found, we return founded value, otherwise we store the value, with given expire time before process the mongo access.

**The storage key is build automatically by our redis implementation**

To see your data from a GUI tool download [Medis](https://github.com/luin/medis)

#### Redis on custom methods

To use redis on custom method just extend your config file like this :

```javascript
{
  "model" : {
    "name" : "Notify",
     .....
    "validator" : "testValidator",
    "fn"        : [
      {
        "type"  : "static", // create a static method on schema
        "name"  : "test1",
        "redis"   : { // HERE YOUR REDIS PART
          "enable"  : true, // MUST BE SET TO TRUE TO ENABLE THIS FUNCTIONALITY
          "expire"  : 10, // EXPIRE TIME FOR STORAGE IN SECONDS
        }
      },
      {
        "type"  : "method", // create a instance method on schema
        "name"  : "test2"
      }
  }
}
```

After that your can access to the redis instance in your custom function and use `add`, `get`, `delete` or `flush` method.

An alias to the add method is provide by the `set` method.
An alias to the delete method is provide by the `remove` method.


```javascript
// valid account schema
exports.test1 = function(data) {
  var redis   = this.redis().instance;
  var expire  = this.redis().expire;

  // add a new key/value
  redis.add('aaa', { foo : 'abar' }, expire);
  // Or
  redis.set('aaa', { foo : 'abar' }, expire);

  // get value
  redis.get('aaa').then(function (success) {
    // do stuff
  }).catch(function (error) {
    // do another stuff here
  });

  // delete an value
  redis.delete('aaa');
  // Or
  redis.remove('aaa');

  // flush values
  redis.flush().then(function (success) {
    // do stuff
  })
  // flush values with a custom pattern
  redis.flush('MY_PATTERN').then(function (success) {
    // do stuff
  })
};
```


CHANGELOG.md
## 2.2.1 (2016-12-01)

- Update README content and add documentation from jsDoc

## 2.2.0 (2016-12-01)

- Add Mongoose Types available on current model

## 2.1.1 (2016-12-01)

- Fix find usage on lodash for enums process
- Update package
- redis.get('YOUR_KEY') method reject the current promise if no key was found
- Update unit tests for delete/remove action on redis usage 
- Change load method to use async.each instead of lodash _.each

## 2.1.0 (2016-11-21)

- Change node engine version to use 4.6.x to use current LTS version until April migration to new lts version
- Add new flush method to provide flush data on current database. This method use pipeline & scan method.
- Update remove method to use pipeline. Now it's possible to send to delete method multiple value for deletion
- Add Mongoose.Schema.Types available on enums modules.
- Add custom pre hook on schema
- Update dependencies modules
- Update node requirement to use 4.6.x LTS Version

## 2.0.0 (2016-10-25)

- Major Update for lodash, joi, ioredis async
- *Remove elasticHosts method. Use enableElasticsearch method instead of elasticHosts.*
- Improve connection manager

## 1.7.2 (2016-06-02)

- Update joi validation to enable host string for Redis and elasticsearch


## 1.7.1 (2016-05-23)

- Bufferize config files when enable ssl authentication for mongodb

## 1.7.0 (2016-05-20)

- Add method remove(key) into redis implementation

## 1.6.0 (2016-05-10)

- Add redis support (add/set & get), on crud method and custom method.

## 1.5.0 (2016-05-10)

- Add SSL/TLS support for elasticsearch connection

## 1.4.6 (2016-04-27)

- Remove the params 'hydrate' when made an esearch() request, now all options will be pass as an options to mongoosastic

## 1.4.5 (2016-04-21)

- Fix [#issue 3](https://github.com/yoctore/yocto-mongoose/issues/3)
- Update Dependencies for mongoose and grunt

## 1.4.4 (2016-04-21)

- Fix [#issue 2](https://github.com/yoctore/yocto-mongoose/issues/2)

## 1.4.3 (2016-04-20)

- Add options on model => [#issue 1](https://github.com/yoctore/yocto-mongoose/issues/1)
- Add rules to authorize auth and protocol properties on default elasticsHosts method

## 1.3.0 (2015-01-16)

- Add middleware post configuration on schema config.
- Add filter on get(s) method
- Add multi update process from upate method

## 1.2.6 (2015-12-09)

- Bug fixe : use 'lodash.extend()' instead of 'lodash.merge()' in the create CRUD method to solve bug with array values

## 1.2.5 (2015-11-10)

- Set the library Q. to handle promise in mongoose

## 1.2.4 (2015-10-12)

- Change content of readme.md
- Change package content

## 1.2.2 & 1.2.3 (2015-10-09)

- Fix a bug on glob.sync on method add function. Now app retrieve correct method file associated with the current method
- Fix join string on glob.sync

## 1.2.1 (2015-09-02)

- Fix bug on alias function
- Add auto exclude of alias when main method is on exclude list

## 1.2.0 (2015-09-02)

- Add alias function for crud methods

## 1.1.0 (2015-09-02)

- Add a new method getValidateSchema when a validator was add.

## 1.0.6 (2015-09-02)

- Add dynamic call of custom function with apply and all given arguments

## 1.0.5 (2015-09-30)

- Fixed logging usage on Method.add

## 1.0.4 (2015-09-30)

- Change usage of merge data on create method to retrieve value

## 1.0.3 (2015-09-30)

- Change regex for method listing
- Add schema instance on given method by dependencies injection

## 1.0.2 (2015-09-30)

- Add instance of enums automaticly on validator function

## 1.0.1 (2015-09-30)

- Change display of king of message when is enums process

## 1.0.0 (2015-09-30)

Add enums process.
- Add Enums class to manage enum action. All schema a by default a static called 'enums' to retrieve enums data by 'get' method

## 0.9.5 (2015-09-29)

- Change logger usage to display debug message on debug mode and all necessary item to info

## 0.9.4 (2015-09-29)

- Fix tests cases for model defintion for crud object

## 0.9.3 (2015-09-29)

- Change logger usage from warning to error for invalid schema during addModel process

## 0.9.2 (2015-09-29)

- Add more test on object property for some cases.

## 0.9.1 (2015-09-29)

- Change path rules for file matching during model load

## 0.9.0 (2015-09-28)

- Change promise middleware from promiseJs to Q

## 0.8.0 (2015-09-28)

- Deleted addStatic method
- Deleted addMethod method
- Deleted addFn method
- Add createMethod method : auto add defined method on current schema from given method directory
- Add a new module "Method". That will automatically add a defined method on a current model model if config json "fn" property was filled.

## 0.7.0 (2015-09-28)

- Generate distribution file on dist directory
- Some fixes

## 0.6.0 (2015-09-28)

General changes :

- Add addStatic method : add a static method to given schema
- Add addMethod method : add a instance method to given schema
- Add addFn method : default function to process add static or instance method on a schema
- Some code fixes
- Add createValidator method : auto add a validator function on create crud if validator name is given on config file

Module changes :

- Add a new validator module to manage auto generation of validate method on create request

## 0.5.0 (2015-09-27)

- Add getModel method to retrieve a valid model

## 0.4.0 (2015-09-27)

- Add addCrud method to enable crud flag from model
- Finish load method for autoloading of models from given path
- Add isLoaded method. This method get model load status

## 0.3.0 (2015-09-27)

- Add get et getOne statics method on crud Class
- Add delete statics method on crud Class

## 0.2.0 (2015-09-27)

- Change controllers Method name to validators

## 0.1.0 (2015-09-25)

Add all base function.
- isConnected : get status of connection
- isDisconnected : get status of disconnection
- connect : connect to database
- disconnect : disconnect if connected on database
- setPath : Min function call during add path action
- isReady : Check if connection is ready

src/modules/utils/redis.js
'use strict';

var logger    = require('yocto-logger');
var _         = require('lodash');
var utils     = require('yocto-utils');
var Redis     = require('ioredis');
var joi       = require('joi');
var Q         = require('q');
var fs        = require('fs');

/**
 *
 * Manage all validator action from a valid given schema
 *
 * @date : 10/04/2016
 * @author : Mathieu ROBERT <mathieu@yocto.re>
 * @copyright : Yocto SAS, All right reserved
 *
 * @class RedisUtils
 */
function RedisUtils (logger) {
  /**
   * Logger instance
   *
   * @property logger
   */
  this.logger     = logger;
  /**
   * Default host to use
   *
   * @type {Array}
   * @default [ { host : '127.0.0.1', port : 6379 } ]
   */
  this.hosts      = [ { host : '127.0.0.1', port : 6379 } ];
  /**
   * Default additional option config
   */
  this.options    = { retry : 2000, enableReadyCheck : true };
  /**
   * Default state of elastic config
   *
   * @type {Boolean}
   * @default false
   */
  this.isReady = false;
  /**
   * Default cluster
   */
  this.cluster = null;
  /**
   * Default expire time
   */
  this.defaultExpireTime = 0;
}

/**
 * Get default hosts for connection
 *
 * @return {Array} list of hosts to use
 */
RedisUtils.prototype.getHosts = function () {
  // default statement
  return this.hosts;
};

/**
 * Get default options for connection
 *
 * @return {Object} list of options to use
 */
RedisUtils.prototype.getOptions = function () {
  // default statement
  return this.options;
};

/**
 * Process redis disconnect action
 */
RedisUtils.prototype.disconnect = function () {
  // process disconnect
  return !_.isNull(this.cluster) ? this.cluster.disconnect() : false;
};

/**
 * Default method to define if hosts is defined
 *
 * @param {Array} hosts an array of hosts given
 * @param {Object} options property to set on options
 * @param {Number} defaultExpireTime to force the default expire time for all of insertion
 * @param {Boolean} cluster set to true if we need to use a cluster connection
 * @return {Boolean} true if all if ok false otherwise
 */
RedisUtils.prototype.connect = function (hosts, options, defaultExpireTime, cluster) {
  // validation schema
  var schema = joi.array().required().items(
    joi.object().keys({
      host      : joi.string().required().empty().default('127.0.0.1'),
      family    : joi.number().optional().valid([ 4, 6 ]),
      password  : joi.string().optional().empty().min(32),
      db        : joi.number().optional(),
      port      : joi.number().required().default(6380),
      tls       : joi.object().optional().keys({
        ca      : joi.string().optional().empty(),
        key     : joi.string().optional().empty(),
        cert    : joi.string().optional().empty(),
      })
    }).default({ host : '127.0.0.1', port : 6380 })
  ).default([ { host : '127.0.0.1', port : 6380 } ]);

  // validate given config
  var validate = joi.validate(hosts, schema);

  // has error ?
  if (validate.error) {
    // log error message
    this.logger.warning([ '[ RedisUtils.connect ] - Invalid host config given :',
                          validate.error ] .join(' '));
    // default invalid statement
    return false;
  }

  // set default expire time
  if (_.isNumber(defaultExpireTime) && defaultExpireTime > 0) {
    // add new value of expire time if is not define on request
    this.defaultExpireTime = defaultExpireTime;
  }

  // set hosts config
  this.hosts    = validate.value;
  // save givent options, this value will be merge before the connection
  this.options  = _.merge(this.options, (options || {}));

  // has an default expire times ?
  if (_.has(this.options, 'retry') && _.isNumber(this.options.retry)) {
    // add retry strategy code
    _.extend(this.options, { clusterRetryStrategy : function () {
      // default statement
      return this.options.retry;
    }.bind(this) });
  }

  // has tls ca defined ?
  if (_.has(this.hosts, 'tls.ca')) {
    // define ca
    this.hosts.tls.ca = [ fs.readFileSync(this.hosts.tls.ca) ];
  }

  // has tls ca defined ?
  if (_.has(this.hosts, 'tls.key')) {
    // define ca
    this.hosts.tls.key = fs.readFileSync(this.hosts.tls.key);
  }

  // has tls ca defined ?
  if (_.has(this.hosts, 'tls.cert')) {
    // define ca
    this.hosts.tls.cert = fs.readFileSync(this.hosts.tls.cert);
  }

  // is cluster mode ?
  if (_.isBoolean(cluster) && cluster) {
    // info message
    this.logger.info('[ RedisUtils.connect ] - Try to connect in a cluster mode');
    // default redis options
    this.cluster = new Redis.Cluster(this.hosts, this.options);
  } else {
    // get default config value
    var params = _.first(validate.value);
    // info message
    this.logger.info('[ RedisUtils.connect ] - Try to connect in classic mode');

    // has more than one items ?
    if (_.size(validate.value) > 1) {
      // warning message
      this.logger.warning([ '[ RedisUtils.connect ] - Redis is not starting on cluster mode',
                            'connections parameters used will be the first item of current config.'
                          ].join(' '));
    }

    // default redis options
    this.cluster = new Redis(params);
  }

  // define event
  return this.listenEvents();
};

/**
 * A default method to normalize given key before save on redis db
 *
 * @param {Mixed} key wanted key for storage
 * @return {String} given key normalized
 */
RedisUtils.prototype.normalizeKey = function (key) {
  // default statement
  return _.snakeCase(_.deburr(JSON.stringify(key))).toLowerCase();
};

/**
 * An alias method for default add method
 *
 * @param {String} key key to use for storage
 * @param {Mixed} value default value to store on database
 * @param {Number} expire expires times to set for current key/value
 * @return {Object} default promise to catch
 */
RedisUtils.prototype.set = function (key, value, expire) {
  // default statement
  return this.add(key, value, expire);
};

/**
 * Add a value from given key on dabatase
 *
 * @param {String} key key to use for storage
 * @param {Mixed} value default value to store on database
 * @param {Number} expire expires times to set for current key/value
 * @return {Object} default promise to catch
 */
RedisUtils.prototype.add = function (key, value, expire) {
  // normalize expire time
  expire = _.isNumber(expire) && expire > 0 ? expire : this.defaultExpireTime;

  // normalize key
  key = this.normalizeKey(key);

  // is ready ?
  if (this.isReady) {
    // expire is a valid time
    if (expire > 0) {
      // log message
      this.logger.debug([ '[ RedisUtils.add ] - Adding key [',
        key, '] with [', expire, '] seconds of expires times,',
        'with value :', utils.obj.inspect(value) ].join(' '));
      // add with expire time
      this.cluster.set(key, JSON.stringify(value), 'EX', expire);
    } else {
      // log message
      this.logger.debug([ '[ RedisUtils.add ] - Adding key [',
        key, '] with value :', utils.obj.inspect(value) ].join(' '));
      // add without expire time
      this.cluster.set(key, JSON.stringify(value));
    }
  }

  // default statement
  return this.get(key);
};

/**
 * Remove a value found by his key
 *
 * @return {boolean} return success status of deleting
 */
RedisUtils.prototype.remove = function () {
  // create async process
  var deferred = Q.defer();

  // build properly all given keys and normalize it
  var keys = _.flatten(_.map(arguments));

  // is ready ?
  if (this.isReady) {
    // create a pipeline to process multiple delete
    var pipeline = this.cluster.pipeline();
    // parse all keys
    keys.forEach(function (key) {
      // prepare deletion
      pipeline.del(this.normalizeKey(key));
    }.bind(this));
    // exec pipeline
    pipeline.exec().then(function () {
      // log message
      this.logger.debug([ '[ RedisUtils.remove ] - Remove keys',
        utils.obj.inspect(keys), 'was processed' ].join(' '));
      // resolve promise
      deferred.resolve();
    }.bind(this));
  } else {
    // log message
    this.logger.warning([ '[ RedisUtils.remove ] - keys [',
      keys, '] was not removed because connection is down' ].join(' '));
  }

  // default statement
  return deferred.promise;
};

/**
 * Flush current database. If a given string pattern is a correct we use it
 *
 * @param {String} pattern if is defined we use this pattern otherwise than default pattern (*)
 * @return {Object} promise to catch
 */
RedisUtils.prototype.flush = function (pattern) {
  // create async process
  var deferred = Q.defer();

  // normalize properly pattern to use
  pattern  = _.isString(pattern) && !_.isEmpty(pattern) ? pattern : '*';

  // current size length, we use it to show a correct message on end stream/pipeline process.
  var sizes = 0;

  // create stream
  var stream = this.cluster.scanStream({
    match : pattern
  });

  // log message
  this.logger.debug('[ RedisUtils.flush ] - flush was starting.');

  // catch data on created stream
  stream.on('data', function (keys) {
    // change sizes valud for end process
    sizes += _.size(keys);
    // has items ?
    if (_.size(keys) > 0) {
      // process call with default method
      this.remove.call(this, keys);
    }
  }.bind(this));

  // when is finish what we do ?
  stream.on('end', function () {
    // no size ???
    if (sizes === 0) {
      // log message
      this.logger.debug([ '[ RedisUtils.flush ] - No keys found for given pattern [',
        pattern, (pattern === '*' ? '(Default) ]' : ']')].join(' '));
    }
    // log message
    this.logger.debug([ '[ RedisUtils.flush ] - flush was ending.',
      sizes, 'keys was removed' ].join(' '));
    // all is ok
    deferred.resolve();
  }.bind(this));

  // default statement
  return deferred.promise;
};

/**
 * An alias method for default remove method
 *
 * @return {Object} default promise to catch
 */
RedisUtils.prototype.delete = function () {
  // default statement
  return this.remove(_.flatten(_.map(arguments)));
};

/**
 * Get current value from given key
 *
 * @param {Mixed} key key to retreive on database
 * @return {Object} default promise to catch
 */
RedisUtils.prototype.get = function (key) {
  // create async process
  var deferred = Q.defer();

  // is ready ?
  if (!this.isReady) {
    // reject
    deferred.reject('Redis is not ready cannot process get action');
  } else {
    // get value
    this.cluster.get(this.normalizeKey(key), function (error, result) {
      // has error ?
      if (!error) {
        // result is null ?
        if (!_.isNull(result)) {
          // resolve with result
          deferred.resolve(JSON.parse(result));
        } else {
          // reject with null value
          deferred.reject(key);
        }
      } else {
        // log error
        this.logger.error([ '[ RedisUtils.get ] - Cannot get value for key [',
          this.normalizeKey(key), '] on redis :', error ].join(' '));
        // reject with error
        deferred.reject(error);
      }
    }.bind(this));
  }
  // default statement
  return deferred.promise;
};

/**
 * Default method to enable listen on each redis events
 *
 * @return {Boolean} return true when event was binded
 */
RedisUtils.prototype.listenEvents = function () {
  // default events
  var events = [
    { key : 'connect', message      : 'is connecting' },
    { key : 'ready', message        : 'is ready' },
    { key : 'error', message        : 'errored' },
    { key : 'close', message        : 'is closing' },
    { key : 'reconnecting', message : 'is reconnecting' },
    { key : 'end', message          : 'is ending' },
    { key : '+node', message        : 'connecting to a new node' },
    { key : '-node', message        : 'disconnecting from a node' },
    { key : 'node error', message   : 'is on error on a node' }
  ];

  // parse all events
  _.each(events, function (e) {
    // catch all event
    this.cluster.on(e.key, function (message) {
      // is ready ?
      if (e.key === 'ready') {
        // change ready state value
        this.isReady = true;
      }
      // error event ? to change ready state
      if (_.includes([ 'error', 'close', 'reconnecting', 'end', 'node error' ], e.key)) {
        // change ready state value
        this.isReady = false;
      }
      // log message
      this.logger[ e.key !== 'ready' ? 'debug' : 'info' ]([
        '[ RedisUtils.listenEvents ] – Redis', e.message,
        (message ? [ '=>', message ].join(' ') : '')
      ].join(' '));
    }.bind(this));
  }.bind(this));

  // default statement
  return true;
};

// Default export
module.exports = function (l) {
  // is a valid logger ?
  if (_.isUndefined(l) || _.isNull(l)) {
    logger.warning('[ RedisUtils.constructor ] - Invalid logger given. Use internal logger');
    // assign
    l = logger;
  }
  // default statement
  return new (RedisUtils)(l);
};
